credentials/tls/certprovider/store.go

```diff
@@ -79,42 +83,107 @@ func GetProvider(name string, config interface{}) (Provider, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	sk := storeKey{
+	pk := providerKey{
 		name:   name,
 		config: string(stableConfig.Canonical()),
 	}
-	if wp, ok := provStore.providers[sk]; ok {
-		wp.refCount++
-		return wp, nil
+
+	// Create a new provider for the (name+config) combo or use an existing one.
+	var wp *wrappedProvider
+	wp, ok := provStore.providers[pk]
+	if !ok {
+		provider := builder.Build(stableConfig)
+		if provider == nil {
+			return nil, fmt.Errorf("certprovider.Build(%v) failed", pk)
+		}
+		wp = &wrappedProvider{
+			prov:     provider,
+			storeKey: pk,
+			store:    provStore,
+		}
+		provStore.providers[pk] = wp
 	}
 
-	provider := builder.Build(stableConfig)
-	if provider == nil {
-		return nil, fmt.Errorf("certprovider.Build(%v) failed", sk)
+	// Create a new distributor for the (name+config+opts) combo or use an
+	// existing one. The reference count for the provider is incremented only
+	// when creating a new distributor.
+	dk := distributorKey{
+		name:    name,
+		config:  string(stableConfig.Canonical()),
+		options: opts,
 	}
-	wp := &wrappedProvider{
-		Provider: provider,
-		refCount: 1,
-		storeKey: sk,
-		store:    provStore,
+	if wd, ok := provStore.distributors[dk]; ok {
+		wd.refCount++
+		return wd, nil
 	}
-	provStore.providers[sk] = wp
-	return wp, nil
+	distributor, err := wp.prov.KeyMaterialReader(opts)
+	if err != nil {
+		return nil, err
+	}
+	wd := &wrappedDistributor{
+		KeyMaterialReader: distributor,
+		refCount:          1,
+		wp:                wp,
+		storeKey:          dk,
+		store:             provStore,
+	}
+	provStore.distributors[dk] = wd
+	wp.refCount++
+	return wd, nil
 }
 
-// Close overrides the Close method of the embedded provider. It releases the
-// reference held by the caller on the underlying provider and if the
-// provider's reference count reaches zero, it is removed from the store, and
-// its Close method is also invoked.
-func (wp *wrappedProvider) Close() {
-	ps := wp.store
-	ps.mu.Lock()
-	defer ps.mu.Unlock()
+// wrappedProvider wraps a provider instance with a reference count.
+type wrappedProvider struct {
+	prov Provider
+	// refcount tracks the number of KeyMaterialReader instances (or
+	// distributors) doled out by this provider.
+	refCount int
+	// A reference to the key and store are also kept here to invoke the
+	// Close method on the provider, and to remove it from the store map.
+	storeKey providerKey
+	store    *store
+}
 
+// close releases the reference held by the caller on the underlying provider
+// and if the provider's reference count reaches zero, it is removed from the
+// store, and its Close method is also invoked.
+//
+// Caller must hold store mutex.
+func (wp *wrappedProvider) close() {
+	ps := wp.store
 	wp.refCount--
 	if wp.refCount == 0 {
-		wp.Provider.Close()
+		wp.prov.Close()
 		delete(ps.providers, wp.storeKey)
```

Comment:
What if someone is holding a ref to a distributor owned by a provider that was closed?  Is this illegal?  (If so, is that documented?)

Reply:
`wrappedProvider.close()` is only called from `wrappedDistributor.Close()` when the underlying `Distributor's` reference count has reached zero. And the `wrappedProvider` maintains a `refCount` of the number of `Distributor's` created on it, and the underlying `Provider.Close()` is only called when each of those `Distributor's` is closed.